oci/archive/oci_src.go

```diff
@@ -13,19 +13,27 @@ import (
 )
 
 type ociArchiveImageSource struct {
-	ref         ociArchiveReference
-	unpackedSrc types.ImageSource
-	tempDirRef  tempDirOCIRef
+	ref              ociArchiveReference
+	unpackedSrc      types.ImageSource
+	tempDirRef       tempDirOCIRef
+	archiveReaderRef *tempDirOCIRef
 }
 
 // newImageSource returns an ImageSource for reading from an existing directory.
 // newImageSource untars the file and saves it in a temp directory
-func newImageSource(ctx context.Context, sys *types.SystemContext, ref ociArchiveReference) (types.ImageSource, error) {
-	tempDirRef, err := createUntarTempDir(sys, ref)
-	if err != nil {
-		return nil, errors.Wrap(err, "creating temp directory")
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref ociArchiveReference, archiveReaderRef *tempDirOCIRef) (types.ImageSource, error) {
+	var (
+		tempDirRef tempDirOCIRef
+		err        error
+	)
+	if archiveReaderRef != nil {
+		tempDirRef = *archiveReaderRef
+	} else {
+		tempDirRef, err = createUntarTempDir(sys, ref)
```

Comment:
Structurally, I’d expect `Reader.List()[…].NewImageSource()` to be the “general case”, with a single `ParseReference().NewImageSource()` constructing a single-use `Reader` here.

The current implementation with `tempDirOCIRef` does amount to a similar effect, but it’s not quite as direct.

Reply:
Fixed